File: internal/server/version_test.go
// SPDX-License-Identifier: Apache-2.0

package server_test

import (
	"net/http"
	"testing"

	"github.com/gopherium/gophenberg/internal/server"
)

type versionBody struct {
	Version string `json:"version"`
}

func TestVersionEndpoint(t *testing.T) {
	t.Parallel()

	users := newFakeUserStore()
	addAda(t, users)
	srv := server.NewServer(server.Config{Users: users, Version: "9.9.9"})

	if code := doRequest(t, srv, http.MethodGet, "/api/version", "").Code; code != http.StatusUnauthorized {
		t.Errorf("unauthenticated GET /api/version = %d, want %d", code, http.StatusUnauthorized)
	}

	cookie := loginCookie(t, srv)
	authed := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		r.AddCookie(cookie)
		srv.ServeHTTP(w, r)
	})
	recorder := doRequest(t, authed, http.MethodGet, "/api/version", "")

	if recorder.Code != http.StatusOK {
		t.Fatalf("status = %d, want %d: %s", recorder.Code, http.StatusOK, recorder.Body.String())
	}
	if body := decodeBody[versionBody](t, recorder); body.Version != "9.9.9" {
		t.Errorf("version = %q, want %q", body.Version, "9.9.9")
	}
}

File: frontend/src/test/version.test.ts
// SPDX-License-Identifier: Apache-2.0

import { http, HttpResponse, server } from '@gophenberg/frontend-sdk/testing'
import { expect, test } from 'vitest'

import { fetchVersion } from '../version'

test('returns the version reported by the backend', async () => {
	server.use(
		http.get('/api/version', () => HttpResponse.json({ version: '1.2.3' })),
	)

	expect(await fetchVersion()).toBe('1.2.3')
})

test('rejects when the version request fails', async () => {
	server.use(
		http.get('/api/version', () =>
			HttpResponse.json({ error: 'internal error' }, { status: 500 }),
		),
	)

	await expect(fetchVersion()).rejects.toThrow('status 500')
})
